### PR TITLE
Attempt to get Save as PDF working

### PR DESCRIPTION
**Pull request by mraible**
See https://github.com/shower/shower/blob/master/docs/printing-en.md for how this should work. Implementation based on https://github.com/opendevise/presentation-bespoke-starter/pull/4.

### CODE REVIEW

**Comment by mojavelinux** [comment]
Here are the steps to enable print styles.

1. Get the file print.styl from https://github.com/opendevise/presentation-bespoke-starter/blob/master/src/styles/print.styl and put it in src/styles.
2. Adjust the width and height in the `@page` selector to match your presentation width and height (e.g., `size 640px 360px`)
3. Add the line `require "./print"` to the *bottom* of main.styl
4. Build the presentation and print! 

You might want to declare variables for your presentation width and height so that you don't have to repeat them in both main.styl and print.styl.

File: presentation/src/styles/main.styl [hunk 1]
```diff
@@ -1,5 +1,6 @@
 @import './fonts.css'
 @import 'normalizecss/normalize.css'
+@require './print'
```
Comment (mojavelinux) [inline presentation/src/styles/main.styl]: This line has to be at the bottom of the stylesheet.

File: presentation/src/styles/print.styl [hunk 2]
```diff
@@ -0,0 +1,29 @@
+$slide-width = 1280px
+$slide-height = 720px
```
Comment (mojavelinux) [inline presentation/src/styles/print.styl]: The width and height need to match the width and height you set in main.styl.

**Comment by mraible** [comment]
This is starting to look pretty good and I dig @hasalex's "with notes" enhancement!

Here's the issues I see for the printable PDF after testing this in Chrome 54 on Mac. I'm guessing these are all fixable with some CSS in `print.styl`.

Slide 5 - why is image pushed down the page? The content of this slide is pretty simple:
    
```
image::conways-law.png[Conway's Law, 300, align=center]
```
	
Slide 9 - why does the quote have an extra line break in the middle? The code is:

```
[quote, Martin Fowler, March 2014]
You shouldn't start with a microservices architecture. Instead begin with a monolith, keep it modular, and split it into microservices once the monolith becomes a problem.
```
	
Slides 10-12 - same issue as slide 5.

Slides 18 & 45 - why is title squished to top and images pushed to bottom?

**Comment by mojavelinux** [comment]
The problem is you are using the wrong print stylesheet. I see a lot of differences between the one you are using and the one at https://github.com/opendevise/presentation-bespoke-starter/blob/master/src/styles/print.styl. That should fix all (or most) of your issues.

Also, you can write the print.styl as follows:

```
@require "./variables"

@page
  size $slide-width $slide-height
  margin 0

...
```

**Comment by mojavelinux** [comment]
The notes stylesheet is awesome. Certainly stuff to refine, but a great start!

**Comment by mraible** [comment]
Thanks @mojavelinux. I switched to the `print.styl` you recommended and it did solve some issues. Now there's only two remaining issues:

* Slides with only an image now have the image top-aligned instead of middle-aligned. Slides 10-12 are good examples.
* Slide 18 with title and image is not aligned in the middle vertically. This doesn't happen on slide 45, but maybe that's because the image is bigger?

Like Keynote, it might work better if the slide title is always at the same position at the top and bullets always start from the same position. This is a minor issue though since I can probably change this in `main.styl` with CSS.

It'd also be cool if the YouTube videos were clickable in the PDF. Will removing the "bespoke-multimedia" plugin make that possible?

**Comment by mojavelinux** [comment]
> It'd also be cool if the YouTube videos were clickable in the PDF. Will removing the "bespoke-multimedia" plugin make that possible?

You'll probably want to have some sort of poster image that is a link. That would require a slightly different behavior in the converter. (I now do this in Asciidoctor PDF). Maybe that is possible with print styles, I'm not sure. You might be able to do something fancy with a data attribute combined with a before/after pseudo-selector. Certainly that would be a nice feature to have.

**Comment by mojavelinux** [comment]
It looks like we're down to working around browser bugs at this point.

For starters, it appears we need to give Chrome a little help in some cases by forcing a page break between slides:

```
@media print
  .bespoke-slide
    page-break-after always
```

That fixes all the slides except 18. It seems like the two figures are repelling each other...resulting in a lot of space that is forcing the title to the previous slide and the java logo to the next slide. I was able to address it by using the following slide content:

```
== Spring Boot Demo

image::spring-boot-logo.png[Spring Boot Demo,100]

{sp}

image::java-logo.png[Java,50]
```

But I can't explain why Chrome is doing this. Just seems like a bug to me.

**Comment by mojavelinux** [comment]
Here's proof slide 18 is demonstrating a flexbox (print) bug. Just group the elements as follows:

```
== !
[.text-center]
--
[discrete]
== Spring Boot Demo
image::spring-boot-logo.png[Spring Boot Demo,150]
image::java-logo.png[Java,50]
--
```

Then add the following style rule to main.styl

```
.text-center
  text-align center
```

Now it works perfectly.

**Comment by mojavelinux** [comment]
My solution to the video link would be as follows. We'd add a link, perhaps even containing a poster image, inside the figure (above the iframe) and hide it using CSS by default. In the print styles, we'd do something like this:

```
.bespoke-slide
  figure.video
    position relative
    a.poster
      display block
      position absolute
      width 100%
      height 100%
```

That way, the link would cover the video and the PDF would present it as an active/clickable region. I'm open to other suggestions.

**Comment by mojavelinux** [comment]
Actually, this should be something perhaps bespoke-multimedia does and the converter just adds a data-href to the figure. That would be cleaner, I think.

**Comment by mraible** [comment]
@mojavelinux I like the idea of doing something in bespoke-multimedia to replace a video with a clickable link in the PDF. However, this plugin makes videos auto-play, which is annoying, so I removed it. 

I tried adding the following to get the images to vertically align:

```
@media print
  .bespoke-slide
    page-break-after always
```

I swear it was working for a minute or two, but now I can't get it to work again.

**Comment by mraible** [comment]
@mojavelinux I created a workaround for video links by adding a screenshot with a link below it that only shows up in the PDF. It'd be cool if I could put the link on the image, but I haven't figured out how to do that.

As far as printing images middle-aligned, I found out that they are centered once I make the browser window smaller by opening the developer console. I had to reduce the small even more to get slide 45 to show up correctly with a title and image. Using `page-break-after always` doesn't seem to matter, it's the size of the browser window that makes the difference.

**Comment by mojavelinux** [comment]
> It'd be cool if I could put the link on the image, but I haven't figured out how to do that.

```
image::poster.jpg[link=https://google.com]
```

However, the Bespoke converter is currently missing this feature. I've filed an issue. See https://github.com/asciidoctor/asciidoctor-bespoke/issues/7.

**Comment by mojavelinux** [comment]
> Using page-break-after always doesn't seem to matter, it's the size of the browser window that makes the difference.

I think it depends. Sometimes it works correctly, sometimes it doesn't. There are clearly some bugs in this view. It comes down to how Chrome is figuring out when a new page is needed...and that seems a bit flaky at the moment.

**Comment by mojavelinux** [comment]
> I tried adding the following to get the images to vertically align:

It's important to understand that the problem is not that the images are not aligning. It's that the space between the images is becoming so large that the content is overflowing. You can see what is happening by adding `overflow: visible` to `.bespoke-slide`. It will show that the slides are bleeding into each other. The content is still vertically aligned properly..it is just starting too soon and ending too late.

Although we see different variations, the root cause seems to be that the spacing between the flexbox items is growing more than it should. It is calculating the boundaries of the box incorrectly.

**Comment by mojavelinux** [comment]
I still think it is good to have the `page-break-after always` declaration since that is the intent. It certainly doesn't seem to hurt and maybe it helps in some cases.

**Pull request merged**